### PR TITLE
Optimize and fill operation about `git stash`

### PR DESCRIPTION
**Pull request by YaaMe**
[#284](https://github.com/k88hudson/git-flight-rules/issues/284)

I notice this command is already listed in another part.
So i fix it and add something new and add `zh-CN` translate.

### CODE REVIEW

**Comment by RichardLitt** [comment]
I wasn't able to make the `stash pop` stuff work.

```sh
(base) 09:11 ~/src/sandbox/sandbox (standardjs) * 🦉 git k
  master
* standardjs
origin	git@github.com:RichardLitt/sandbox.git (fetch)
origin	git@github.com:RichardLitt/sandbox.git (push)
On branch standardjs
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	new file:   test-2

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	test

(base) 09:11 ~/src/sandbox/sandbox (standardjs) * 🦉 git stash
Saved working directory and index state WIP on standardjs: 1998329 function
(base) 09:11 ~/src/sandbox/sandbox (standardjs) * 🦉 git k
  master
* standardjs
origin	git@github.com:RichardLitt/sandbox.git (fetch)
origin	git@github.com:RichardLitt/sandbox.git (push)
On branch standardjs
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	test

nothing added to commit but untracked files present (use "git add" to track)
(base) 09:11 ~/src/sandbox/sandbox (standardjs) * 🦉 git stash pop --index 1
error: patch failed: README.md:1
error: README.md: patch does not apply
Conflicts in index. Try without --index.
```

File: README.md [hunk 1]
```diff
@@ -1466,6 +1469,22 @@ $ git stash apply "stash@{n}"
 
 Here, 'n' indicates the position of the stash in the stack. The topmost stash will be position 0.
 
+Furthermore, time based aslo possible(If you can remeber).
```
Comment (RichardLitt) [inline README.md]: Spelling

**Comment by YaaMe** [comment]
typo fixed. And...

It is because you have conflicts between current status & stash.So you have to manually handle these conflicts....I think this action depends on yourself..`add untracked files` or `reset` or decide to keep which part of it.
[more detail](https://stackoverflow.com/questions/19003009/how-to-recover-stashed-uncommitted-changes?answertab=active#tab-top)
`git reset --hard` will lost your untracked files without any hints.So i think show conflicts here will be better.If you think it's a conflict idea i can split this commit.
~~Actually,`conflict` is your git friend who loyally tell you something you missed, why everyone are scared of him?~~

**Comment by RichardLitt** [comment]
Thanks @YaaMe. Just to test it out, I used a fresh git repo. Here are my results, which still aren't satisfactory: 

```sh
  1 # First Edit!
(base) 16:08 ~/src/sandbox/git-stash 🦉 git init
Initialized empty Git repository in /Users/richard/src/sandbox/git-stash/.git/
(base) 16:08 ~/src/sandbox/git-stash 🦉 touch README.md && git ca -m 'Init'
[master (root-commit) 9539035] Init
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 README.md
(base) 16:08 ~/src/sandbox/git-stash (master) 🦉 git k
* master
On branch master
nothing to commit, working tree clean
(base) 16:08 ~/src/sandbox/git-stash (master) 🦉 vi README.md
(base) 16:08 ~/src/sandbox/git-stash (master) * 🦉 touch testfile
(base) 16:08 ~/src/sandbox/git-stash (master) * 🦉 git add README.md
(base) 16:08 ~/src/sandbox/git-stash (master) * 🦉 git k
* master
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   README.md

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	testfile

(base) 16:08 ~/src/sandbox/git-stash (master) * 🦉 echo "Now, I want to stage my unstaged edits, and unstage my staged edits - so, stage testfile, and unstage README.""
> "
Now, I want to stage my unstaged edits, and unstage my staged edits - so, stage testfile, and unstage README.

(base) 16:09 ~/src/sandbox/git-stash (master) * 🦉 git stash
Saved working directory and index state WIP on master: 9539035 Init
(base) 16:09 ~/src/sandbox/git-stash (master) * 🦉 git k
* master
On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	testfile

nothing added to commit but untracked files present (use "git add" to track)
(base) 16:09 ~/src/sandbox/git-stash (master) * 🦉 git stash pop --index 1
refs/stash@{1} is not a valid reference
(base) 16:09 ~/src/sandbox/git-stash (master) * 🦉 git k
* master
On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	testfile

nothing added to commit but untracked files present (use "git add" to track)
(base) 16:09 ~/src/sandbox/git-stash (master) * 🦉 echo "Stash pop did not work."
Stash pop did not work.
```

**Comment by YaaMe** [comment]
Untracked is not unstaged...
I handle it in `magit`, so i always pick single file and won't get any problem..That's reason i suggest it..
I'm not so sure what do you want to do....
For this case
Assume you want to unstage README and to stage testfile
```sh
$ git reset HEAD README.md
$ git add testfile
```
Assume here's multi files and you don't want to control one by one.
```sh
$ git reset HEAD # unstage files
$ git stash # hide them
$ git add . # add untracked files
$ git stash pop # give back
```

And I pretty sure
```sh
$ git stash -k
$ git reset --hard
$ git stash pop # These are `git reset HEAD`
$ git add -A # and you will stage all files after exec this command
```

If you want to untracked README, it's pretty wired without a commit **remove(files)**,`git rm --cached -r .` will help and you will get a conflict when `git stash pop` because they're already in commit....

**Comment by RichardLitt** [comment]
I want to do this: "### I want to stage my unstaged edits, and unstage my staged edits"

That's what's in the section that you've changed in this PR, which is why the current PR doesn't work. Perhaps you meant to change a different part of the document?

**Comment by YaaMe** [comment]
I think it's a fix now....by following the doc I staged all edits...

File: README.md [hunk 2]
```diff
@@ -605,13 +606,15 @@ Then, you will need to use the `e` option to manually choose which lines to add.
 <a href="unstaging-edits-and-staging-the-unstaged"></a>
 ### I want to stage my unstaged edits, and unstage my staged edits
 
-This is tricky. The best I figure is that you should stash your unstaged edits. Then, reset. After that, pop your stashed edits back, and add them.
+In many cases, You should unstage all files and then pick file you want and commit it.
+But assume that's what you want,You can create a temp commit to store your stage files, stage your unstaged files and stash them, Then, reset the temp commit and stash pop.
 
 ```sh
-$ git stash -k
```
Comment (RichardLitt) [inline README.md]: We should add a note that git stash or git stash k doesn't work with newfiles, only with edits.

File: README.md [hunk 3]
```diff
@@ -605,13 +606,15 @@ Then, you will need to use the `e` option to manually choose which lines to add.
...
-$ git reset --hard
-$ git stash pop
-$ git add -A
+$ git commit -m "WIP"
+$ git add . # also have untracked files
+$ git stash
+$ git reset HEAD^
```
Comment (RichardLitt) [inline README.md]: Why Head^ over --hard?

File: README.md [hunk 4]
```diff
@@ -605,13 +606,15 @@ Then, you will need to use the `e` option to manually choose which lines to add.
...
-This is tricky. The best I figure is that you should stash your unstaged edits. Then, reset. After that, pop your stashed edits back, and add them.
+In many cases, You should unstage all files and then pick file you want and commit it.
+But assume that's what you want,You can create a temp commit to store your stage files, stage your unstaged files and stash them, Then, reset the temp commit and stash pop.
 
 ```sh
-$ git stash -k
-$ git reset --hard
-$ git stash pop
-$ git add -A
+$ git commit -m "WIP"
```
Comment (RichardLitt) [inline README.md]: I would leave this line out. I don't know what it adds.

File: README.md [hunk 5]
```diff
@@ -605,13 +606,15 @@ Then, you will need to use the `e` option to manually choose which lines to add.
...
+$ git add . # also have untracked files
```
Comment (RichardLitt) [inline README.md]: Same for this. Let's assume that the user already has staged and unstaged edits.

File: README.md [hunk 6]
```diff
@@ -605,13 +606,15 @@ Then, you will need to use the `e` option to manually choose which lines to add.
...
+$ git stash
+$ git reset HEAD^
+$ git stash pop --index 0
```
Comment (RichardLitt) [inline README.md]: What does `--index 0` do that `git stash pop` alone doesn't?

File: README.md [hunk 7]
```diff
@@ -1448,6 +1451,10 @@ $ git stash push working-directory-path/filename1.ext working-directory-path/fil
 ```sh
 $ git stash save <message>
 ```
+or
+```sh
+$ git push -m <message>
```
Comment (RichardLitt) [inline README.md]: Do you mean `git stash -m message`?

File: README.md (hunk 2 above)
Comment (YaaMe) [reply README.md]: you may need `-u` option

File: README.md (hunk 3 above)
Comment (YaaMe) [reply README.md]: `--hard` is involved HEAD, index and files. I don't want to lost my files.Only rollback the last commit marked "WIP" to get my staged files.

File: README.md (hunk 5 above)
Comment (YaaMe) [reply README.md]: You can create a new local repo to try this.
I mark staged files m1 m2 unstaged files n1 n2 and untracked file n3.
[this link](https://stackoverflow.com/questions/20028507/git-stash-uncached-how-to-put-away-all-unstaged-changes/34681302#34681302) maybe help you.
I make a temp commit to store staged files.And then add unstaged files.
Actually, I suggest to handle it one by one...

File: README.md (hunk 6 above)
Comment (YaaMe) [reply README.md]: I think it's a bug..not sure it's a bug of `magit` or `git` or my mind...I find that `git stash pop` make all stash unstaged. and `git stash pop --index 0` keep it staged.Maybe i could google it....

File: README.md (hunk 7 above)
Comment (YaaMe) [reply README.md]: sry, `git stash push -m <message>` i mean,

File: README.md (hunk 2 above)
Comment (RichardLitt) [reply README.md]: Can you add something for that?

File: README.md (hunk 3 above)
Comment (RichardLitt) [reply README.md]: Ah, I see what you're doing now. That makes sense, actually, and is easier. Maybe we could keep both ways in this document?

File: README.md (hunk 6 above)
Comment (RichardLitt) [reply README.md]: Could you check?

File: README.md (hunk 3 above)
Comment (YaaMe) [reply README.md]: I could, but the old way is wrong.You will mark everything staged.

File: README.md (hunk 6 above)
Comment (YaaMe) [reply README.md]: [here](https://stackoverflow.com/questions/31595873/git-stash-with-staged-files-does-stash-convert-staged-files-to-unstaged?answertab=active#tab-top)
I will add a note here.

File: README.md (hunk 2 above)
Comment (YaaMe) [reply README.md]: I think It's already [here](https://github.com/k88hudson/git-flight-rules#stash-all-edits) :p

**Comment by RichardLitt** [comment]
@YaaMe I've edited the PR (see 40a542d31a85f2edf69e87a70a841161775c0907) and then merged it. Thank you for this effort! :)

**Pull request merged**